crates/sui-indexer/src/handlers/checkpoint_handler.rs

```diff
@@ -1,1174 +1,255 @@
+                    warn!("Failed to persist first epoch, retrying...");
+                    persist_first_epoch_res = self.state.persist_epoch(&epoch).await;
                 }
+                epoch_db_guard.stop_and_record();
+                self.metrics.total_epoch_committed.inc();
+                info!("Persisted genesis epoch");
+            } else {
+                // NOTE: when the channel is full, epoch_sender_guard will wait until the channel has space.
+                self.epoch_indexing_sender.send(epoch).await.map_err(|e| {
+                    error!(
+                        "Failed to send indexed epoch to epoch commit handler with error {}",
+                        e.to_string()
+                    );
+                    IndexerError::MpscChannelError(e.to_string())
+                })?;
             }
-            let seq = checkpoint.checkpoint.sequence_number;
-            info!(
-                checkpoint_seq = seq,
-                elapsed, "Checkpoint indexing finished, about to sending to commit handler"
-            );
-            // NOTE: when the channel is full, checkpoint_sender_guard will wait until the channel has space.
-            // Checkpoints are sent sequentially to stick to the order of checkpoint sequence numbers.
-            self.checkpoint_sender
-                .send(checkpoint)
-                .await
-                .tap_ok(|_| info!(checkpoint_seq = seq, "Checkpoint sent to commit handler"))
-                .unwrap_or_else(|e| {
-                    panic!(
-                        "checkpoint channel send should not fail, but got error: {:?}",
-                        e
-                    )
-                });
         }
+        let seq = checkpoint.checkpoint.sequence_number;
+        info!(
+            checkpoint_seq = seq,
+            elapsed, "Checkpoint indexing finished, about to sending to commit handler"
+        );
+        // NOTE: when the channel is full, checkpoint_sender_guard will wait until the channel has space.
+        // Checkpoints are sent sequentially to stick to the order of checkpoint sequence numbers.
+        self.checkpoint_sender
+            .send(checkpoint)
+            .await
+            .tap_ok(|_| info!(checkpoint_seq = seq, "Checkpoint sent to commit handler"))
+            .unwrap_or_else(|e| {
+                panic!(
+                    "checkpoint channel send should not fail, but got error: {:?}",
+                    e
+                )
+            });
+
+        Ok(())
     }
+}
 
-    async fn index_checkpoint_and_epoch(
-        state: &S,
-        data: &CheckpointTxData,
-    ) -> Result<(TemporaryCheckpointStore, Option<TemporaryEpochStore>), IndexerError> {
-        let CheckpointTxData {
-            checkpoint,
-            transactions,
-            system_state_objects: _,
-        } = data;
+struct CheckpointDataObjectStore<'a> {
+    objects: &'a [sui_types::object::Object],
```

Comment:
Given we're only looking up this info in checkpoints which contain epoch change txns we should be good since those are usually very small checkpoints.